### ISSUE TITLE
Improve file-based kernel cache

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#4926]
* Often `~/.cupy/kernel_cache` inode becomes very big. I think it's better to introduce a hierarchy e.g. using the first 2 chars from the cache key: `~/.cupy/ab/abXXXXXXXXX....cubin`.
* Avoid appending MD5 to a file. The reason this is introduced was to support Windows that needs special care to do atomic rename (https://github.com/cupy/cupy/pull/37/files).
  `os.replace` is in Python 3 so we can just use it now.
  (FYI, on Windows `os.replace` uses `MoveFileExW` with `MOVEFILE_REPLACE_EXISTING` flag which can be considered atomic in our usecase; there is a long thread in the [forum](https://social.microsoft.com/Forums/partner/ja-JP/449bb49d-8acc-48dc-a46f-0760ceddbfc3/movefileexmovefilereplaceexisting-ntfs-same-volume-atomic?forum=windowssdk).)
* Include some hint info to filename to manage cache files easily. (maybe NVRTC version and compute capability)

**Comment by leofang** [comment]
> * Often `~/.cupy/kernel_cache` inode becomes very big.

I've been wondering about this. I would be interested in exploring solutions like periodically invalidating outdated caches (any update of CUDA or CuPy or CC versions could trigger this). We could also do it a bit aggressively, say purging caches older than a month. So far I've been cleaning this manually, and it would be great if there's a mechanism doing a better job than II do...😄 

The hierarchy approach is not mutually exclusive to this auto invalidation, nor does it address the inode inflation problem entirely.

**Comment by kmaehashi** [comment]
How about adding this tool https://github.com/cupy/cupy/blob/master/.pfnci/trim_cupy_kernel_cache.py to `cupyx/tools`?

> Include some hint info to filename to manage cache files easily. (maybe NVRTC version and compute capability)

Also including a CuPy's major version ('8', '9', etc.) to filename may help managing caches.

**Comment by leofang** [comment]
> How about adding this tool https://github.com/cupy/cupy/blob/master/.pfnci/trim_cupy_kernel_cache.py to `cupyx/tools`?

Yeah, something like this would likely help. Do users need to run this manually? Is it cross platform (i.e. support Windows)?

**Comment by kmaehashi** [comment]
> Do users need to run this manually?

Yes, users need to manually run the tool. Once we have a hint tag in filename we can add some more convenient filters (e.g. purge all caches for non-current CUDA versions).
I'm not sure about running it automatically. If we do this for example at startup users may see it as "`import cupy` is extremely slow.
Maybe ading a warning to run the tool at shutdown if `os.stat('.cupy/kernel_cache').st_size` (which is O(1) = do not have to  iterate all cache files) exceeds some threshold value?

> Is it cross platform (i.e. support Windows)?

Yeah I wrote this for Windows CI 😄

**Comment by leofang** [comment]
> Maybe ading a warning to run the tool at shutdown if `os.stat('.cupy/kernel_cache').st_size` (which is O(1) = do not have to iterate all cache files) exceeds some threshold value?

Sounds good to me, but now I am concerned with a multi-process environment like Dask or mpi4py. We might need a lock to do this...?

> > Is it cross platform (i.e. support Windows)?
> 
> Yeah I wrote this for Windows CI 😄

Right, I forgot, sorry 😅

**Comment by leofang** [comment]
btw it'd be nice to also distinguish from the hint tag whether it's ptx being cached or cubin (#4591).

**Comment by leofang** [comment]
related? #1890
